### DISCUSSION TITLE
Allow to define JSON encoding options in Eloquent API Resources

### DISCUSSION THREAD

**Discussion by Gummibeer** [laravel/framework#38136] [Ideas] [answered]
PHP supports several options on how to encode JSON - some of the most important ones for APIs could be `JSON_PRESERVE_ZERO_FRACTION ` or `JSON_UNESCAPED_UNICODE`. But it's not possible to pass these options to the JSON encoding point from within an Eloquent API Resource.

`\Illuminate\Http\Resources\Json\JsonResource::toResponse()` 
https://github.com/laravel/framework/blob/6e9baa07af67f2ae44127ef7e93762f3fd42868f/src/Illuminate/Http/Resources/Json/JsonResource.php#L214-L223

`\Illuminate\Http\Resources\Json\ResourceResponse::toResponse()`
https://github.com/laravel/framework/blob/6e9baa07af67f2ae44127ef7e93762f3fd42868f/src/Illuminate/Http/Resources/Json/ResourceResponse.php#L29-L49

**Idea 1 - new method to define options**
The idea to solve this would be a new (optional) method `jsonEncodingOptions()` on the resource class itself which is used in the `ResourceResponse` class and passed to the `response()->json()` method as you would do in your own controller.
The downside of that would be that one resource can contain another resource and they could define different options but only the top one used to generate the response would be respected. This could produce unexpected behavior for the users if not properly communicated.

**Idea 2 - resolve `ResourceResponse` via container**
My second idea would be to resolve the `ResourceResponse` class via the container in the `JsonResource` class which would allow the user to override and rebind that single class. Would define one set of options for the whole application and doesn't require the user to override and extend classes only to be able to add custom behavior to one single method.

-----

In theory that is also a bigger "problem" as it could also be helpful to define application-wide encoding options used by default instead of `0` everywhere. So something like a `config('app.json_encoding_options', 0)`. Which is used by every class that implements some kind of `toJson()` method.
As the default wouldn't change none of these changes should be breaking - so it could be added to every version but because of the massive optional behavior change it could be useful to add these things to Laravel 9 - only to stay safe.

**Comment by mikemand** [comment]
The second part of your Idea 2 would be fantastic (some kind of global configuration option that sets the same JSON options throughout the application). I currently manually call `->toJson()` on a lot of stuff, because I want PHP to throw on any JSON error rather than outputting malformed JSON. I also pass `JSON_NUMERIC_CHECK`, because 9 times out of 10 I'm working with the JSON in JavaScript and would rather have _some_ typing rather than everything as strings. You can't have floating-point errors with strings, after all. Where's the fun in that?

**Reply by Gummibeer** [reply]
Hey,
yes - this would be a very cool addition but also a much larger refactoring that's why I've only added it but would say that it's a bit out of scope of that idea/issue. To clarify it I've added a divider in my original post.

But I totally agree that an overall JSON configuration would solve many problems. Would be similar like other settings we already have - language used for translations and carbon for example.
So I think that having one option to define the overall application/framework JSON behavior would be great.

**Reply by Gummibeer** [reply]
Only for completeness: one default set of JSON options could also get problematic as you would want to use different options to write JSON to files than for your API responses.
So keeping them separated and allowing/having the user to override several classes and define the options there would be more work for the user but fewer problems overall.

**Comment by nope123nope** [comment]
This would be a very cool addition for those, who use resource transformers as a platform for json response.
For me, passing UGC html througt api (utf-8 encoded as default) it's to heavy. 
Upvoted this discussion @Gummibeer thanks for mentioning this approach

**Comment by tripleten** [comment, marked as answer]
Don't we already have a method to set the encoding options in the API Resource?

You can do it by adding the below method either in your API collection or resource class:

```
/**
     * Customize the outgoing response for the resource.
     *
     * @param  \Illuminate\Http\Request
     * @param  \Illuminate\Http\Response
     * @return void
     */
    public function withResponse($request, $response)
    {   
        $response->header('Charset', 'utf-8');
        $response->header('Any-other-header', 'header-value');
        $response->setEncodingOptions(JSON_UNESCAPED_UNICODE);
    }
```

Thoughts, @Gummibeer @mikemand & @nope123nope ?

**Reply by NicoHood** [reply]
This works great for me!

**Reply by NicoHood** [reply]
Why isnt that documented in the docs?

**Reply by NicoHood** [reply]
[I've improved it a bit:](https://stackoverflow.com/a/72411129/14348748)

```php
/**
 * Customize the outgoing response for the resource.
 *
 * @param  \Illuminate\Http\Request
 * @param  \Illuminate\Http\Response
 * @return void
 */
public function withResponse($request, $response)
{
    $charset = $request->header('Accept-Charset', 'utf-8');
    if ($charset === 'utf-8') {
        $response->header('Charset', 'utf-8');
        $response->header('Content-Type','application/json; charset=utf-8');
        $response->setEncodingOptions(JSON_UNESCAPED_UNICODE);
    }
    // Alternative: iso-8859-1 which will be encoded by default using json escaping.
}
```
